app/models/manageiq/providers/cloud_manager/orchestration_template_runner.rb

```diff
@@ -0,0 +1,165 @@
+    save!
+    service.add_resource(@orchestration_stack)
+    my_signal(false, :poll_stack_status, 10)
+  rescue StandardError => err
+    _log.log_backtrace(err)
+    my_signal(minimize_indirect, :post_stack_run, err.message, 'error')
+  ensure
+    # create options may never be saved before unless they were overridden
+    service.save_create_options
+  end
+
+  def update_orchestration_stack
+    set_status('updating orchestration stack')
+
+    update_options = options.except(:service_id)
+    # use orchestration_template from service_template, which may be different from existing orchestration_template
+    service.orchestration_stack.raw_update_stack(service.service_template.orchestration_template, update_options)
+    self.name = "#{name}, update Orchestration Stack ID: #{service.orchestration_stack.id}"
+    miq_task.update_attributes(:name => name)
+    save!
+    my_signal(false, :poll_stack_status, 10)
+  rescue StandardError => err
+    _log.log_backtrace(err)
+    my_signal(minimize_indirect, :post_stack_run, err.message, 'error')
+  end
+
+  def poll_stack_status(interval)
+    set_status('checking orchestration stack deployment status')
+
+    status, message = service.orchestration_stack ? service.orchestration_stack.raw_status.normalized_status : ["check_status_failed", "stack has not been deployed"]
+    options.merge!(:orchestration_stack_status => status, :orchestration_stack_message => message)
+    save!
+    _log.info("Stack deployment status: #{status}, reason: #{message}")
+
+    case status.downcase
+    when 'create_complete', 'update_complete'
+      my_signal(minimize_indirect, :post_stack_run, "Orchestration stack [#{service.stack_name}] #{status}", 'ok')
+    when 'rollback_complete', 'delete_complete', /failed$/, /canceled$/
+      _log.error("Orchestration stack deployment error: #{reason}. Please examine stack resources for more details")
+      my_signal(minimize_indirect, :post_stack_run, "Orchestration stack deployment error: #{reason}", 'error')
+    else
+      interval = 60 if interval > 60
+      my_signal(false, :poll_stack_status, interval * 2, :deliver_on => Time.now.utc + interval)
+    end
+  rescue MiqException::MiqOrchestrationStackNotExistError, MiqException::MiqOrchestrationStatusError => err
+    # naming convention requires status to end with "failed"
+    options.merge!(:orchestration_stack_status => 'check_status_failed', :orchestration_stack_message => err.message)
+    save!
+    _log.log_backtrace(err)
+    my_signal(minimize_indirect, :post_stack_run, err.message, 'error')
+  end
+
+  def post_stack_run(message, status)
+    my_signal(true, :finish, message, status)
+  end
+
+  def set_status(message, status = "ok")
+    _log.info(message)
+    super
+  end
+
+  def my_signal(no_queue, action, *args, deliver_on: nil)
+    if no_queue
+      signal(action, *args)
+    else
+      queue_signal(action, *args, :deliver_on => deliver_on)
+    end
+  end
+
+  def service
```

Comment:
Try our best to decouple modules. Avoid `service` in this class. The job options should contain only parameters necessary to deploy a stack. For update the options can include stack_id too.